kivy/tests/test_lang_pre_process_and_post_process.py

```diff
@@ -0,0 +1,176 @@
+        '''
+        testcase = self
+        ae = self.assertEqual
+        NP = NumericProperty
+
+        class EventCounter(Factory.Widget):
+            does_have_root_rule = BooleanProperty(True)
+            _n_pre_from_d = NP(0)  # 'd' stands for 'default handler'
+            _n_post_from_r = NP(0)  # 'r' stands for 'root rule'
+            _n_post_from_c = NP(0)  # 'c' stands for 'class rule'
+            _n_post_from_d = NP(0)
+
+            def on_kv_pre(self):
+                self._n_pre_from_d += 1
+                ae(self._n_pre_from_d, 1)
+                ae(self._n_post_from_r, 0)
+                ae(self._n_post_from_c, 0)
+                ae(self._n_post_from_d, 0)
+
+            def on_kv_pre_from_c(self):
+                testcase.fail('This method is not supposed to be called')
+
+            def on_kv_pre_from_r(self):
+                testcase.fail('This method is not supposed to be called')
+
+            def on_kv_post(self):
+                self._n_post_from_d += 1
+                self.assert_all_handlers_were_called_correctly()
+
+            def on_kv_post_from_c(self):
+                self._n_post_from_c += 1
+                ae(self._n_pre_from_d, 1)
+                ae(self._n_post_from_r, 1 if self.does_have_root_rule else 0)
+                ae(self._n_post_from_c, 1)
+                ae(self._n_post_from_d, 0)
+
+            def on_kv_post_from_r(self):
+                if not self.does_have_root_rule:
+                    testcase.fail(
+                        "Strange. The handler was called even though "
+                        "the widget doesn't have root rule.")
+                self._n_post_from_r += 1
+                ae(self._n_pre_from_d, 1)
+                ae(self._n_post_from_r, 1)
+                ae(self._n_post_from_c, 0)
+                ae(self._n_post_from_d, 0)
+
+            def assert_all_handlers_were_called_correctly(self):
+                ae(self._n_pre_from_d, 1)
+                ae(self._n_post_from_r, 1 if self.does_have_root_rule else 0)
+                ae(self._n_post_from_c, 1)
+                ae(self._n_post_from_d, 1)
+
+        # Test #1: Without root rule
+        root = EventCounter(does_have_root_rule=False)
+        root.assert_all_handlers_were_called_correctly()
+
+        # Test #2: With root rule
+        root = Builder.load_string(textwrap.dedent('''
+        EventCounter:
+            on_kv_pre: self.on_kv_pre_from_r()  # won't be excuted
+            on_kv_post: self.on_kv_post_from_r()
+            EventCounter:
+                id: child
+                on_kv_pre: self.on_kv_pre_from_r()  # won't be excuted
+                on_kv_post: self.on_kv_post_from_r()
+        '''))
+        root.assert_all_handlers_were_called_correctly()
+        root.ids.child.assert_all_handlers_were_called_correctly()
+
+        # Test #3: If the user add a widget during 'on_kv_pre()',
+        #          is 'on_kv_post' still dispatched once on that widget?
```

Comment:
`Test #3` is the reason why I added a new property `_on_kv_post_was_done`. Without the property, `on_kv_post` is fired twice on `self._ec`. So yeah it's not a good design.